src/gfn/samplers.py

```diff
@@ -264,3 +268,275 @@ def sample_trajectories(
+                back_steps > trajectories.when_is_done,
+                trajectories.when_is_done,
+                back_steps,
+            )
+
+        backward_trajectories = self.backward_sampler.sample_trajectories(
+            env,
+            states=trajectories.last_states,
+            conditioning=conditioning,
+            save_estimator_outputs=save_estimator_outputs,
+            save_logprobs=save_logprobs,
+            **policy_kwargs,
+        )
+
+        # Calculate the forward probability if needed (Metropolis-Hastings).
+        prev_trajectories = Trajectories.reverse_backward_trajectories(
+            backward_trajectories
+        )
+        prev_trajectories_log_rewards = trajectories.log_rewards
+
+        all_states = backward_trajectories.to_states()
+        junction_states = all_states[torch.arange(bs, device=device) + bs * K]
+
+        ### Reconstructing with self.estimator
+        recon_trajectories = super().sample_trajectories(
+            env,
+            states=junction_states,
+            conditioning=conditioning,
+            save_estimator_outputs=save_estimator_outputs,
+            save_logprobs=save_logprobs,
+            **policy_kwargs,
+        )
+
+        # Obtain full trajectories by concatenating the backward and forward parts.
+        new_trajectories_dones = (
+            backward_trajectories.when_is_done - K + recon_trajectories.when_is_done
+        )
+        new_trajectories_log_rewards = recon_trajectories.log_rewards  # Episodic reward
+
+        max_traj_len = new_trajectories_dones.max() + 1
+        new_trajectories_states_tsr = torch.full(
+            (max_traj_len, bs, *state_shape), -1
+        ).to(trajectories.states.tensor)
+        new_trajectories_actions_tsr = torch.full(
+            (max_traj_len - 1, bs, *action_shape), -1
+        ).to(trajectories.actions.tensor)
+
+        # Calculate the log probabilities as needed.
+        if save_logprobs:
+            log_pf_prev_trajectories = get_trajectory_pfs(
+                pf=self.estimator, trajectories=prev_trajectories
+            )
+            log_pf_recon_trajectories = get_trajectory_pfs(
+                pf=self.estimator, trajectories=recon_trajectories
+            )
+            log_pf_new_trajectories = torch.full((max_traj_len - 1, bs), 0.0).to(
+                device=device, dtype=torch.float
+            )
+        if use_metropolis_hastings:
+            log_pb_prev_trajectories = get_trajectory_pbs(
+                pb=self.backward_sampler.estimator,
+                trajectories=prev_trajectories,
+            )
+            log_pb_recon_trajectories = get_trajectory_pbs(
+                pb=self.backward_sampler.estimator, trajectories=recon_trajectories
+            )
+            log_pb_new_trajectories = torch.full((max_traj_len - 1, bs), 0.0).to(
+                device=device, dtype=torch.float
+            )
+
+        for i in range(bs):  # FIXME: Can we vectorize this?
```

Comment:
I think this should be vectorized -- if you need help, no problem, let's schedule a pair programming session (maybe in the evening EST instead of my early morning ;) ).

Reply:
Let me check first! I will let you know :)